[DataPipe] Enabling graph traversal for MapDataPipe

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #74851

This PR allows `MapDataPipe` to be traced as part of the graph traversal function. This is necessary because some `IterDataPipes` such as converter or `MapKeyZipper` can take `MapDataPipe` as inputs.

NivekT committed:
[DataPipe] Enabling graph traversal for MapDataPipe

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enabling graph traversal for MapDataPipe"


Will add another PR related to DataPipe serialization issues that are raised.

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enabling graph traversal for MapDataPipe"


This PR allows `MapDataPipe` to be traced as part of the graph traversal function. This is necessary because some `IterDataPipes` such as converter or `MapKeyZipper` can take `MapDataPipe` as inputs.

[ghstack-poisoned]

ejguan commented on test/test_datapipe.py:
```diff
@@ -2008,6 +2008,20 @@ def test_traverse_forked(self):
+        graph = torch.utils.data.graph.traverse(iter_dp)
+        expected: Dict[Any, Any] = {iter_dp: {source_map_dp: {}}}
+        self.assertEqual(expected, graph)
+
```
After this PR is landed, do you mind add the above test into torchdata test case?

ejguan commented on torch/utils/data/graph.py:
```diff
@@ -39,33 +39,37 @@ def reduce_hook(obj):
             captured_connections.append(obj)
             return _stub_unpickler, ()
 
+    datapipe_classes: Set[Type[DataPipe]] = {IterDataPipe, MapDataPipe}
```
Super nit: List or tuple should be fine here. Set has a little bit overhead.

ejguan reviewed (approved):
Thank you for adding this patch. LGTM

NivekT committed:
Update on "[DataPipe] Enabling graph traversal for MapDataPipe"


This PR allows `MapDataPipe` to be traced as part of the graph traversal function. This is necessary because some `IterDataPipes` such as converter or `MapKeyZipper` can take `MapDataPipe` as inputs.

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enabling graph traversal for MapDataPipe"


This PR allows `MapDataPipe` to be traced as part of the graph traversal function. This is necessary because some `IterDataPipes` such as converter or `MapKeyZipper` can take `MapDataPipe` as inputs.

[ghstack-poisoned]

NivekT commented:
@pytorchbot merge this please

NivekT commented:
@pytorchbot merge this please
